File: 3757.number-of-effective-subsequences.py
#
# @lc app=leetcode id=3757 lang=python3
#
# [3757] Number of Effective Subsequences
#

# @lc code=start
from typing import List

class Solution:
    def countEffective(self, nums: List[int]) -> int:
        MOD = 10**9 + 7
        n = len(nums)
        if n == 0:
            return 0
        BITS = 20
        U = (1 << BITS) - 1
        MASK = 0
        for num in nums:
            MASK |= num
        bits = [b for b in range(BITS) if (MASK & (1 << b))]
        K = len(bits)
        f = [0] * (1 << BITS)
        for num in nums:
            f[num] += 1
        g = f[:]
        for bit in range(BITS):
            p = 1 << bit
            for mask in range(1 << BITS):
                if mask & p:
                    g[mask] += g[mask ^ p]
        ans = 0
        for s in range(1, 1 << K):
            mask_T = 0
            pop = 0
            for j in range(K):
                if s & (1 << j):
                    mask_T |= (1 << bits[j])
                    pop += 1
            zero_cnt = g[U ^ mask_T]
            req_size = n - zero_cnt
            num_subs = pow(2, n - req_size, MOD)
            if pop % 2 == 1:
                ans = (ans + num_subs) % MOD
            else:
                ans = (ans - num_subs + MOD) % MOD
        return ans
# @lc code=end

File: 3757.number-of-effective-subsequences.cpp
#
# @lc app=leetcode id=3757 lang=cpp
#
# [3757] Number of Effective Subsequences
#

# @lc code=start
class Solution {
public:
    int countEffective(vector<int>& nums) {
        const int MOD = 1e9 + 7;
        int n = nums.size();
        
        // Compute full OR
        int full_or = 0;
        for (int x : nums) full_or |= x;
        
        // Find bits in full_or
        vector<int> bits;
        for (int i = 0; i < 20; i++) {
            if (full_or & (1 << i)) bits.push_back(i);
        }
        int k = bits.size();
        
        // Compress nums[i] to indices and count
        vector<int> cnt(1 << k, 0);
        for (int x : nums) {
            int idx = 0;
            for (int i = 0; i < k; i++) {
                if (x & (1 << bits[i])) {
                    idx |= (1 << i);
                }
            }
            cnt[idx]++;
        }
        
        // SOS DP: dp[mask] = sum of cnt[v] for all v that are subsets of mask
        vector<int> dp = cnt;
        for (int i = 0; i < k; i++) {
            for (int mask = 0; mask < (1 << k); mask++) {
                if (mask & (1 << i)) {
                    dp[mask] += dp[mask ^ (1 << i)];
                }
            }
        }
        
        // Precompute powers of 2
        vector<long long> pow2(n + 1);
        pow2[0] = 1;
        for (int i = 1; i <= n; i++) {
            pow2[i] = pow2[i-1] * 2 % MOD;
        }
        
        // Inclusion-exclusion
        long long ans = 0;
        int full_mask = (1 << k) - 1;
        for (int sub = 1; sub <= full_mask; sub++) {
            int complement = full_mask ^ sub;
            int union_cnt = n - dp[complement];
            long long contribution = pow2[n - union_cnt];
            if (__builtin_popcount(sub) % 2 == 1) {
                ans = (ans + contribution) % MOD;
            } else {
                ans = (ans - contribution + MOD) % MOD;
            }
        }
        
        return (int)ans;
    }
};
# @lc code=end